contributors/design-proposals/vertical-pod-autoscaler.md

```diff
@@ -0,0 +1,495 @@
+by a VPA config with mode not set to “off”, the controller rewrites the request
+by applying recommended resources to the Pod spec. Otherwise it leaves the Pod
+spec unchanged.
+
+The controller gets the recommended resources by fetching
+/recommendedPodResources from the Recommender. If the call times out or fails,
+the controller falls back to the recommendation cached in the VPA object.
+If this is also not available and the Pod doesn’t have a VPA Initializer, it is
+rejected.
+
+### Life of the Pod ###
+
+Once Initializers are introduced, there will be two paths for a Pod using VPA to
+get scheduled.
+
+1. A synchronous (fast) path: if a matching VPA exists at the time of creating
+   the Pod and it is able to obtain a recommendation (either by calling the
+   Recommender or by using a cached value in the VPA resource), the Admission
+   Plugin sets resources synchronously. It also immediately deletes the VPA
+   initializer if one was present, to shorten the critical path for Pod
+   creation.
+
+2. An asynchronous (slower) path. A Pod is created with the Initializer.
+   The Admission Plugin doesn't see a matching VPA so it passes the Pod
+   unchanged. Once a matching VPA appears, the VPA controller sets the Pod
+   resources and removes the Initializer (one could also consider adding
+   optional deadline to the Initializer to let the Pod schedule if the VPA
+   doesn’t appear).
+
+![VPA Scheduling Flow](images/vpa-scheduling-flow.png "VPA scheduling flow")
+
+### Recommender ###
+Recommender is the main component of the VPA. It is responsible for
+computing recommended resources. On startup the recommender fetches
+historical resource utilization of all Pods (regardless of whether
+they use VPA) together with the history of Pod OOM events from the
+History Storage. It aggregates this data and keeps it in memory.
+
+During normal operation the recommender consumes real time updates of
+resource utilization and new events via the Master Metrics API from
+Heapster. Additionally it watches all Pods and all VPA objects in the
+cluster. For every Pod that is matched by some VPA selector the
+Recommender computes the recommended resources and sets the
+recommendation on the VPA object.
+
+It is important to realize that one VPA object has one recommendation.
+The user is expected to use one VPA to control Pods with similar
+resource usage patterns, typically a group of replicas or shards of
+a single workload.
+
+The Recommender exposes a synchronous RPC interface that takes a Pod
+Spec and the Pod metadata and returns recommended resources:
+
+```json
+GET /recommendedPodResources
+{
+  "metadata": { },
+  "spec": { }
+}
+```
+
+The response:
+
+```json
+{
+  "containers": [
+    {"name": "...", "resources": { }}
+  ]
+}
+```
+
+Notice that this API may be called for an existing Pod, as well as for a
+yet-to-be-created Pod.
+It is not guaranteed that this API will return the same recommendation for all
```

Comment:
Good catch. I was thinking of situations where VPA will need to adjust the recommendation before applying it to a specific pod, based on the current state of the cluster (things like quota, space available on nodes or other scheduling constraints). But we don't really have a solution to this yet and it requires more research, both in terms of desired semantics and possible implementation.
Still, whatever solution we come up with to this problem, it will most likely be implemented in the Updater and not the Recommender. So I guess this sentence should be removed.